### PR TITLE
General Applied Math Functions

### LINKED ISSUE #76: General Applied Math Functions
A few general functions for working with applied math

- [abs](https://www.ncl.ucar.edu/Document/Functions/Built-in/abs.shtml)
- [avg](https://www.ncl.ucar.edu/Document/Functions/Built-in/avg.shtml)
- [cumsum](https://www.ncl.ucar.edu/Document/Functions/Built-in/cumsum.shtml)
- [decimalPlaces](https://www.ncl.ucar.edu/Document/Functions/Contributed/decimalPlaces.shtml)
- [get_d2r](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_d2r.shtml)
- [get_r2d](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_r2d.shtml)
- [get_pi](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_pi.shtml)
- [log](https://www.ncl.ucar.edu/Document/Functions/Built-in/log.shtml)
- [log10](https://www.ncl.ucar.edu/Document/Functions/Built-in/log10.shtml)
- [mod](https://www.ncl.ucar.edu/Document/Functions/Contributed/mod.shtml)
- [product](https://www.ncl.ucar.edu/Document/Functions/Built-in/product.shtml)
- [sum](https://www.ncl.ucar.edu/Document/Functions/Built-in/sum.shtml)

### PR DESCRIPTION
**Pull request by cyschneck**
## PR Summary
Simple general applied math functions

- [abs](https://www.ncl.ucar.edu/Document/Functions/Built-in/abs.shtml)
- [avg](https://www.ncl.ucar.edu/Document/Functions/Built-in/avg.shtml)
- [cumsum](https://www.ncl.ucar.edu/Document/Functions/Built-in/cumsum.shtml)
- [decimalPlaces](https://www.ncl.ucar.edu/Document/Functions/Contributed/decimalPlaces.shtml)
- [get_d2r](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_d2r.shtml)
- [get_r2d](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_r2d.shtml)
- [get_pi](https://www.ncl.ucar.edu/Document/Functions/Contributed/get_pi.shtml)
- [log](https://www.ncl.ucar.edu/Document/Functions/Built-in/log.shtml)
- [log10](https://www.ncl.ucar.edu/Document/Functions/Built-in/log10.shtml)
- [mod](https://www.ncl.ucar.edu/Document/Functions/Contributed/mod.shtml)
- [product](https://www.ncl.ucar.edu/Document/Functions/Built-in/product.shtml)
- [sum](https://www.ncl.ucar.edu/Document/Functions/Built-in/sum.shtml)

## Related Tickets & Documents
Closes #76

## PR Checklist
**General**
- [x] PR includes a summary of changes
- [x] Link relevant issues, make one if none exist

### CODE REVIEW

**Comment by cyschneck** [comment]
Some smaller functions that I bundled into the same notebook, while still doing research for larger functions

**Comment by jukent** [comment]
This is great @cyschneck ! Do you plan on having an application page as well for this?

**Comment by cyschneck** [comment]
Would it be useful to make one? These functions are short enough to put together into a single ncl_entry notebook

**Comment by jukent** [comment]
> Would it be useful to make one? These functions are short enough to put together into a single ncl_entry notebook

I'm not sure, I was thinking just for find-ability - seeing math functions from the main applications page.

**Comment by cyschneck** [comment]
Agreed, I'll write up a Python Applications to add to the main page

File: applications/data_analysis/general_applied_math.ipynb
Comment (kafitzgerald) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>I worry a little bit about the effort required to create and maintain a resource like this for all basic functionality.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (kafitzgerald) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>Should <code>math</code> be referred to as a module rather than a package?</p><p> </p><p>We also probably should mention here that the numpy functions are typically much faster and optimized to work on array data (you do get to this second part a bit).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>More of a meta level question for discussion here...</p><p> </p><p>I wonder a little bit about the value of spelling out the individual examples here for all of these functions.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>We might put something in here about what these are e.g. "This section covers general applied math functions from NCL" just to orient the reader.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>It's a bit repetitive, but I wonder if doing something like <a href="https://ncar.github.io/geocat-applications/ncl/ncl_entries/days_in_month.html#overview" target="_blank">what was done for days_in_month</a> might be good here too e.g. "NCL's <code>abs</code> returns..." just so it's clear this is the NCL function. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    abs_output = abs(input_value)
</pre><p> </p><p>Is there a reason not to use the numpy version (np.abs or np.absolute)? Seems more consistent with other stuff, but I might be missing something.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    round_output = [round(value, 3) for value in input_values]
</pre><p>same comment here as with abs.</p><p> </p><p>and the other non-numpy functions below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>I'm not sure this is the case for rounding / truncating. It looks like the behavior is a little different between NCL and Python (standard lib and numpy), but there's a little more nuance to it:</p><p><a href="https://docs.python.org/3/library/functions.html#round" target="_blank">https://docs.python.org/3/library/functions.html#round</a></p><p><a href="https://numpy.org/doc/stable/reference/generated/numpy.round.html" target="_blank">https://numpy.org/doc/stable/reference/generated/numpy.round.html</a></p><p><a href="https://www.ncl.ucar.edu/Document/Functions/Built-in/round.shtml" target="_blank">https://www.ncl.ucar.edu/Document/Functions/Built-in/round.shtml</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>I think we decided to leave this to another section, right?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:1]: <p>Is there a reason not to use <code>radians</code> and <code>degrees</code> from numpy?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

File: applications/data_analysis/general_applied_math.ipynb
Comment (cyschneck) [reply applications/data_analysis/general_applied_math.ipynb:1]: Do you think the depth for the NCL entry is enough? I can remove this if it is going to make the scale too big

Comment (kafitzgerald) [reply applications/data_analysis/general_applied_math.ipynb:1]: Maybe leave it in for now and let's see what others think about the balance here.  @anissa111 @jukent 

I'm also not opposed to merging this with the general applications entry and feeling it out from there just wanted to raise it as a potential longer term concern for folks to think about.

Thanks for really getting things rolling with content - this is great!

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I think there is value is keeping the individual examples the same level across the ncl_entries, just to keep the pages consistent

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: Ok, maybe let's leave as is for now then and see how things scale.

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: It is a better way to translate the array so I'll change that, but I think not requiring the external package when it is isn't needed for simple math can make it complicated. This notebook uses the standard library when it can for single inputs since it doesn't need any additional work or package wrangling.

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: Same here, I changed the array from a list comprehension to numpy, but kept it the same for single input from the standard library

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: `radians` is the same as `deg2rad`, but `deg2rad` is a more descriptive name and in the [code](https://stackoverflow.com/questions/51801646/difference-between-deg2rad-and-radians-in-numpy) `radians` uses `deg2rad`. I wouldn't be surprised if numpy deprecates `radians`/`degrees` in the future since it is duplicated functions.

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I think we'll probably want to include the numpy versions of functions where available given the audience (lots of arrays and place where you'd want to use a numpy ufunc if possible), but including both probably doesn't hurt.

This is what I was getting at below in a few places as well.

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: Sure, I was more curious about the use of math vs either set of numpy functions (see above).

I'll close this specific comment though in favor of the general discussion above.

Comment (anissa111) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: <p>+1 here.</p><p> </p><p>I also think the language around <code>math.ceil</code> being "rounding up" and <code>math.floor</code> being "rounding down" could be clarified. I personally don't think of ceiling and floor functions as rounding, but I'm interested in hearing others' thoughts here.</p>

File: applications/data_analysis/general_applied_math.ipynb
Comment (anissa111) [reply applications/data_analysis/general_applied_math.ipynb:1]: <p>If I had noticed before you made this content, I would have maybe discouraged making a python-first resource for this in geocat-applications, but since it's already made, I'm also not opposed to merging it in.</p>

Comment (anissa111) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>small thing: I'd consider using an admonition for the "important note" bit here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (anissa111) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>I know that this is (probably) technically accurate, but it feels reductive. I'd prefer if we removed the text under the heading.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (anissa111) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>Reword on "are also available as well as part of the <code>numpy</code> library"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (anissa111) [inline applications/data_analysis/general_applied_math.ipynb:1]: <p>Possible copy/paste error? -&gt; "By default, the input and output values are in radians."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NCAR/geocat-applications/pull/77/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (kafitzgerald) [reply applications/data_analysis/general_applied_math.ipynb:1]: I'll resolve this for now, but let's keep this in mind for future design discussions.

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I'm ok with the "round up" "round down" language.  It is a bit confusing here with so many definitions of round, but I think it's in the Python docs with similar language and I don't have a better suggestion offhand.  
 
I was more concerned that this statement is confusing and/or needs more clarification: 
"NCL distinguishes between rounding a decimal value and truncating it, but Python rounds the values by default, while truncating requires converting to a string" 

And that this statement isn't true in some cases "The Python function round function rounds a float up when 5 or greater and rounds down when 4 or lower."  For both math and numpy rounding is done toward the even choice so this switches back and forth.

It's probably also worth noting that the NCL round function only rounds to the nearest whole number, which is considerably less flexible than the Python implementations.

I do think the differences sections are a little tricky in general and this function in particular is surprisingly quite different in NCL from the Python ones.  

I don't know that this has to be perfect in terms of language and don't want to hold this up too much before merging, but I am worried about the correctness at least.

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: We can log an issue and separate out this conversation, but I feel pretty strongly that we should be at least including examples of functions that operate on arrays.

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I see what you mean. So, NCL function this is mean to cover is [decimalPlaces](https://www.ncl.ucar.edu/Document/Functions/Contributed/decimalPlaces.shtml), which either truncates or rounds the values.  decimalPlaces rounds/truncates a value to a specific number of decimal places

Meanwhile, NCL [round](https://www.ncl.ucar.edu/Document/Functions/Built-in/round.shtml) function rounds the float/double to nearest whole number. The round function isn't currently being covered here, but will be as part of v2 (https://github.com/NCAR/geocat-applications/issues/79). I can clear up the language here though

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I expanded the functions to include working with lists and operating with `numpy` functions that return arrays

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: This looks better overall, but it looks like the description of the Python round function still needs an update.

> And that this statement isn't true in some cases "The Python function round function rounds a float up when 5 or greater and rounds down when 4 or lower." For both math and numpy rounding is done toward the even choice so this switches back and forth.

Comment (kafitzgerald) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: It looks like they're still missing from the NCL Index and the receipt as well.  I think it's worth making the change on the index at least.  The receipt probably isn't a bad idea though and would be more consistent w/ the trig functions.

File: ncl/ncl_index/ncl-index-table.csv
Comment (kafitzgerald) [inline ncl/ncl_index/ncl-index-table.csv:25]: Super minor thing, but it looks like the text in the notes section should probably be lowercase for consistency.

Comment (cyschneck) [reply ncl/ncl_index/ncl-index-table.csv:25]: Sure thing!

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: The NCL index now references `abs() or numpy.abs()`, but I have expanded the receipt to include math and numpy within the v2 https://github.com/NCAR/geocat-applications/issues/79, so I can resolve that with the next PR

Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb:1]: I reworded it a bit and added a note based on a disclaimer in the Python round function

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (kafitzgerald) [inline ncl/ncl_entries/general_applied_math.ipynb:317]:
```suggestion
    "The `round` function returns the value rounded to ndigits after the decimal point\n",
```

**Review by kafitzgerald — COMMENTED** [review]
One more minor suggested change I think we should make here, but otherwise I think this is probably good to get merged.  

There is still some nuance missing in the truncate / round differences section (e.g. generally rounding to the even value in Python and some algorithmic differences between Python's `round` and `np.round`), but that's down in the weeds a bit and some of the numpy functions are still missing from the index and examples.  I think you mentioned this (i.e. including the numpy functions) was addressed in the next PR though.

File: ncl/ncl_entries/general_applied_math.ipynb
Comment (cyschneck) [reply ncl/ncl_entries/general_applied_math.ipynb]: Yes, the additional numpy work is in v2, but I expanded on this to detail the kind of rounding that is happening here with some disclaimers

**Review by kafitzgerald — APPROVED** [review]

**Review by anissa111 — COMMENTED** [review]
Looks good! 

One last thing before I approve, can we get rid of the metadata changes to `datetime.ipynb`? I know they're just metadata differences, but I think it's worth removing the changes to avoid this PR showing up in that file's commit history and causing any confusion in the future.

**Comment by cyschneck** [comment]
Good catch, I removed `datetime.ipynb`

**Review by anissa111 — APPROVED** [review]

**Pull request merged**
